digdag-cli/src/main/java/io/digdag/cli/profile/TaskAnalyzer.java

```diff
@@ -0,0 +1,103 @@
+
+    private final Injector injector;
+
+    private static class AttemptIdWithSiteId
+    {
+        public final int siteId;
+        public final long attemptId;
+
+        public AttemptIdWithSiteId(int siteId, long attemptId)
+        {
+            this.siteId = siteId;
+            this.attemptId = attemptId;
+        }
+    }
+
+    public TaskAnalyzer(Injector injector)
+    {
+        this.injector = injector;
+    }
+
+    private void waitAfterDatabaseAccess(int waitMillis)
+    {
+        try {
+            TimeUnit.MILLISECONDS.sleep(waitMillis);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException(e);
+        }
+    }
+
+    public TasksSummary run(Instant createdFrom,
+                            Instant createdTo,
+                            int fetchedAttempts,
+                            int partitionSize,
+                            int databaseWaitMillis)
+    {
+        TransactionManager tm = injector.getInstance(TransactionManager.class);
+        SessionStoreManager sm = injector.getInstance(SessionStoreManager.class);
+
+        AtomicLong lastId = new AtomicLong();
+        TasksSummary.Builder tasksSummaryBuilder = new TasksSummary.Builder();
+        logger.debug("Task analysis started");
+
+        while (true) {
+            logger.debug("Collecting attempts");
+
+            // Partition target attempts to avoid a too long transaction later
+            List<List<AttemptIdWithSiteId>> attemptIdsList = tm.begin(() ->
+                    Lists.partition(
+                            sm.findFinishedAttemptsWithSessions(createdFrom, createdTo, lastId.get(), fetchedAttempts).stream()
+                                    .map(attempt -> {
+                                        lastId.set(Math.max(lastId.get(), attempt.getId()));
+                                        return new AttemptIdWithSiteId(attempt.getSiteId(), attempt.getId());
+                                    })
+                                    .collect(Collectors.toList()),
+                            partitionSize));
+
+            logger.debug("Collected {} attempts", attemptIdsList.stream().mapToInt(List::size).sum());
+            waitAfterDatabaseAccess(databaseWaitMillis);
+
+            if (attemptIdsList.isEmpty()) {
+                break;
+            }
+
+            // Process partitioned attemptIds list from the beginning
+            for (List<AttemptIdWithSiteId> attemptIds : attemptIdsList) {
+                logger.debug("Processing {} attempts", attemptIds.size());
+                tm.begin(() -> {
+                    for (AttemptIdWithSiteId attemptIdWithSiteId : attemptIds) {
+                        List<ArchivedTask> tasks = sm.getSessionStore(attemptIdWithSiteId.siteId).getTasksOfAttempt(attemptIdWithSiteId.attemptId);
+                        TasksSummary.updateBuilderWithTasks(tasksSummaryBuilder, tasks);
```

Comment:
Is there any reason that `updateBuilderWithTasks` is static method of `TaskSummary` ?
I guess this is  a job for `TaskSummaryBuilder`. So  `TaskSummaryBuilder.updateWithTasks(tasks)` may be better.
But not strong opinion.

Reply:
Good point! Fixed it